enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,181 @@
+
+**System Administrator** is a human user responsible for managing Edge Devices.
+When a system administrator triggers a software or operating system upgrade
+on Edge Devices based on `rpm-ostree`, the changes do not take place until the
+next system reboot.
+
+**End User** is a human user operating an Edge Device. End users continue their
+work with the device running the previous software version until the reboot.
+
+Once the system is rebooted implicitly (i.e. `cron` schedule) or explicitly
+by a system administrator, 
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+[![Mermaid Upgrade Flow](https://mermaid.ink/img/pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR?type=png)](https://mermaid.live/edit#pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR)
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed. 
+
+The MicroShift core RPM package will include a health check script to validate
+that all the required MicroShift services are up and running. The script will
+be installed into `/etc/greenboot/check/required.d` directory conditional to
+the presence of the `greenboot` package. This would prevent unnecessary scripts
+from being installed on systems that are not `rpm-ostree` based.
+
+The health check script will implement the following validations. Exiting the
+script with a non-zero status will have the boot declared as failed.
+
+|Validation                                          |Pass  |Fail  |
+|----------------------------------------------------|------|------|
+|microshift.service is enabled                       |Next  |exit 0|
+|Wait for any pods to start                          |Next  |exit 1|
+|Wait for `openshift-ovn-kubernetes` pods to be ready|Next  |exit 1|
```

Comment:
About the other parallel proposal: https://github.com/openshift/enhancements/pull/1304 (effort of making that modular/optional)

To avoid coupling with a specific CNI, I propose we don't wait for ovn-kubernetes; the other services (ca, ingress, DNS) will only boot once networking is ready, so it becomes implicit that whatever the networking was, it became ready.

Another alternative is thinking of this in a way that will be modular (if you install an RPM, the check for that component goes on; otherwise, it's off).

Reply:
We also have storage which will be pluggable eventually, so we would need to account for that as well.
Once this specific functionality is developed, we will need to update the script to make the list of cheked namespaces conditional. 
I will update the proposal.